pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -824,71 +830,103 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
+					klog.Errorf("Failed while getting a Node to retry updating node health. Probably Node %s was deleted.", name)
+					return false, err
 				}
+				return false, nil
+			}); err != nil {
+				klog.Errorf("Update health of Node '%v' from Controller error: %v. "+
+					"Skipping - no pods will be evicted.", node.Name, err)
+				return
+			}
+
+			// Some nodes may be excluded from disruption checking
+			if !isNodeExcludedFromDisruptionChecks(node) {
+				zoneToNodeConditionsLock.Lock()
+				zoneToNodeConditions[nodetopology.GetZoneKey(node)] = append(zoneToNodeConditions[nodetopology.GetZoneKey(node)], currentReadyCondition)
+				zoneToNodeConditionsLock.Unlock()
 			}
 
-			_, needsRetry := nc.nodesToRetry.Load(node.Name)
-			switch {
-			case currentReadyCondition.Status != v1.ConditionTrue && observedReadyCondition.Status == v1.ConditionTrue:
-				// Report node event only once when status changed.
-				controllerutil.RecordNodeStatusChange(nc.recorder, node, "NodeNotReady")
-				fallthrough
-			case needsRetry && observedReadyCondition.Status != v1.ConditionTrue:
-				if err = controllerutil.MarkPodsNotReady(ctx, nc.kubeClient, nc.recorder, pods, node.Name); err != nil {
-					utilruntime.HandleError(fmt.Errorf("unable to mark all pods NotReady on node %v: %v; queuing for retry", node.Name, err))
-					nc.nodesToRetry.Store(node.Name, struct{}{})
-					continue
+			if currentReadyCondition != nil {
+				pods, err := nc.getPodsAssignedToNode(node.Name)
+				if err != nil {
+					utilruntime.HandleError(fmt.Errorf("unable to list pods of node %v: %v", node.Name, err))
+					if currentReadyCondition.Status != v1.ConditionTrue && observedReadyCondition.Status == v1.ConditionTrue {
+						// If error happened during node status transition (Ready -> NotReady)
+						// we need to mark node for retry to force MarkPodsNotReady execution
+						// in the next iteration.
+						nc.nodesToRetry.Store(node.Name, struct{}{})
+					}
+					return
+				}
+				if nc.runTaintManager {
+					nc.processTaintBaseEviction(ctx, node, &observedReadyCondition)
+				} else {
+					if err := nc.processNoTaintBaseEviction(ctx, node, &observedReadyCondition, gracePeriod, pods); err != nil {
+						utilruntime.HandleError(fmt.Errorf("unable to evict all pods from node %v: %v; queuing for retry", node.Name, err))
+					}
+				}
+
+				_, needsRetry := nc.nodesToRetry.Load(node.Name)
+				switch {
+				case currentReadyCondition.Status != v1.ConditionTrue && observedReadyCondition.Status == v1.ConditionTrue:
+					// Report node event only once when status changed.
+					controllerutil.RecordNodeStatusChange(nc.recorder, node, "NodeNotReady")
+					fallthrough
+				case needsRetry && observedReadyCondition.Status != v1.ConditionTrue:
+					if err = controllerutil.MarkPodsNotReady(ctx, nc.kubeClient, nc.recorder, pods, node.Name); err != nil {
+						utilruntime.HandleError(fmt.Errorf("unable to mark all pods NotReady on node %v: %v; queuing for retry", node.Name, err))
```

Comment:
nit: Error message starts with Caps to be consistent with the code base.

Reply:
This line isn't actually part of my change, it's just the whitespace changing